server/src/main/java/org/opensearch/common/xcontent/JsonToStringXContentParser.java

```diff
@@ -100,40 +127,76 @@ private void parseToken(StringBuilder path, String currentFieldName) throws IOEx
                     this.keyList.add(fieldNameSuffix);
                 }
             } else if (this.parser.currentToken() == Token.START_ARRAY) {
-                parseToken(path, currentFieldName);
+                lastBrotherValueValid = parseToken(path, currentFieldName, depth);
+                isChildrenValueValid |= lastBrotherValueValid;
                 break;
             } else if (this.parser.currentToken() == Token.END_ARRAY) {
                 // skip
             } else if (this.parser.currentToken() == Token.START_OBJECT) {
-                parseToken(path, currentFieldName);
-                int dotIndex = path.lastIndexOf(DOT_SYMBOL, path.length());
-
-                if (dotIndex != -1 && path.length() > currentFieldName.length()) {
-                    path.setLength(path.length() - currentFieldName.length() - 1);
-                }
+                path.append(DOT_SYMBOL).append(currentFieldName);
+                boolean validValue = parseToken(path, currentFieldName, depth + 1);
+                isChildrenValueValid |= validValue;
+                removeKeyIfNeed(validValue);
+                assert path.length() > (currentFieldName.length() - 1);
+                path.setLength(path.length() - currentFieldName.length() - 1);
             } else {
-                if (!path.toString().contains(currentFieldName)) {
+                String parseValue = parseValue();
+                if (parseValue != null) {
                     path.append(DOT_SYMBOL).append(currentFieldName);
-                }
-                parseValue(parsedFields);
-                this.valueList.add(parsedFields.toString());
-                this.valueAndPathList.add(path + EQUAL_SYMBOL + parsedFields);
-                int dotIndex = path.lastIndexOf(DOT_SYMBOL, path.length());
-                if (dotIndex != -1 && path.length() > currentFieldName.length()) {
+                    this.valueList.add(parseValue);
+                    this.valueAndPathList.add(path + EQUAL_SYMBOL + parseValue);
+                    isChildrenValueValid = true;
+                    lastBrotherValueValid = true;
+                    assert path.length() > (currentFieldName.length() - 1);
                     path.setLength(path.length() - currentFieldName.length() - 1);
+                } else {
+                    lastBrotherValueValid = false;
                 }
             }
+        }
+        if (visitFieldName && isChildrenValueValid == true) {
+            removeKeyIfNeed(lastBrotherValueValid);
+        }
+        return isChildrenValueValid;
+    }
 
+    public void removeKeyIfNeed(boolean hasValidValue) {
```

Comment:
I' m not sure whether we should remove the key has null value, it also be discusses in here: https://github.com/opensearch-project/OpenSearch/issues/13880#issuecomment-2143113169

It will indeed impact the exist query when making change.